library/systemd/src/lib/yast2/system_service.rb

```diff
@@ -175,92 +175,134 @@ def start_mode=(mode)
+    #   active value of the service in the system.
     #
     # @return [Boolean] true if the service must be active; false otherwise
     def active?
       return new_value_for(:active) if changed_value?(:active)
-      service.active?
+      current_active?
+    end
+
+    # Keywords to search for this service
+    #
+    # In case the service has an associated socket, the socket name
+    # is included as keyword.
+    #
+    # @return [Array<String>] e.g., #=> ["tftp.service", "tftp.socket"]
+    def keywords
+      keywords = [service.id]
+      keywords << socket.id if socket
+      keywords
     end
 
-    # Sets the service to be started after calling #save
+    # Sets the service to be started after calling {#save}
     #
     # @see #active=
+    #
+    # @return [Symbol] :start
     def start
       self.active = true
+      self.action = :start
     end
 
-    # Sets the service to be stopped after calling #save
+    # Sets the service to be stopped after calling {#save}
     #
     # @see #active=
+    #
+    # @return [Symbol] :stop
     def stop
       self.active = false
+      self.action = :stop
     end
 
-    # TODO
-    def restart
-    end
-
-    # TODO
-    def reload
-    end
-
-    # TODO
-    #
-    # Action to perform when the service is saved (see {#save})
+    # Sets the service to be restarted after calling {#save}
     #
-    # @return [Symbol, nil] :start, :stop, :restart or :reload. It returns nil if no
-    #   action has been requested or the requested action does not modify the service
-    #   (e.g., to stop a service when the service is not active).
-    def action
+    # @return [Symbol] :restart
+    def restart
+      register_change(:active, true)
+      self.action = :restart
     end
 
-    # TODO: remove this
-    # Toggles the service status
+    # Sets the service to be reloaded after calling {#save}
     #
-    # @see #active=
-    def toggle
-      self.active = !active?
+    # @return [Symbol] :reload
+    def reload
+      register_change(:active, true)
+      self.action = :reload
     end
 
     # Saves changes to the underlying system
     #
-    # @param set_status [Boolean] Do not change service status. Useful when running on 1st stage.
-    def save(ignore_status: false)
+    # @param keep_state [Boolean] Do not change service status. Useful when running on 1st stage.
+    #
+    # @return [Boolean]
+    def save(keep_state: false)
       clear_errors
       save_start_mode
-      set_current_status unless ignore_status
-      reset
+      perform_action unless keep_state
+      refresh
     end
 
-    # Reverts stored changes
+    # Reverts cached changes
+    #
+    # @return [Boolean]
     def reset
       clear_changes
+      @action = nil
+
+      true
+    end
+
+    # Refreshes the underlying system and discards cached changes
+    #
+    # @return [Boolean]
+    def refresh
+      service.refresh!
+      @start_modes = []
       @current_start_mode = nil
+
+      reset
```

Comment:
why refresh do reset? I think main reason to have that refresh is to just recheck current state and recompute if changes still makes sense. But I think that it should not reset actions, not?

Reply:
Yes, it is better to have `#reset` and `#refresh` as completely separate methods and call both inside `#save`.